stan-dev/math
# Cleaning up makefiles.

## Checklist

- [x] Math issue #954

- [x] Copyright holder: Generable

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the code base is stable

- [x] the changes are maintainable


- [x] the changes adhere to the Math library's [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality)


## Summary
This pull request simplifies the build and gets the makefiles in order. The makefiles got tangled up through a few recent changes. It also allows the builds to be more extensible in the Math library.

Here are some of the highlights:
- `CXX` is the only compiler setting used. In develop, both `CXX` and `CC` have to be set
- Windows: removed the `uname: not found` message
- `CXXFLAGS` are now broken into smaller components that can be set by the user:
    - `CXXFLAGS_LANG` for language-specific flags
    - `CXXFLAGS_WARNINGS` for ignoring warnings
    - `CXXFLAGS_BOOST` for boost-specific flags
    - `CXXFLAGS_SUNDIALS` for sundials-specific flags
    - `CXXFLAGS_EIGEN` for eigen-specific flags
    - `CXXFLAGS_GTEST` for google-test-specific flags
    - `CXXFLAGS_OS` for a catch-all and any os-specific flags.
- `CPPFLAGS` are broken into components (`CPPFLAGS` are the C preprocessor flags):
    - `CPPFLAGS_BOOST` for boost-specific flags
    - `CPPFLAGS_SUNDIALS` for sundials-specific flags
    - `CPPFLAGS_EIGEN` for eigen-specific flags
    - `CPPFLAGS_GTEST` for google-test-specific flags
    - `CPPFLAGS_OPENCL` for opencl

  breaking apart the flags this way makes it clearer how to extend the flags for MPI and GPU.
- each of the flags can be set within `make/local` and it will take precedence over the default values
- the dependency generation now generates with dummy targets for each file. This means more robustness in that -- this fixes problems where a header file is removed from the source tree. This caused a lot of trouble with stability of make.
- it also cuts the size of the makefiles in roughly half (while having all the same behavior)
- also fixes #986. Fixes issues with math includes in tests for Windows. Thanks, @rok-cesnovar!

## Tests

Run tests from a clean checkout without setting anything in `make/local`

By that, I mean:
1. A few unit tests. This should test enough things:
```
./runTests.py test/unit/math_include_test.cpp test/unit/multiple_translation_units_test.cpp test/unit/math/rev/mat/functor/cvodes_ode_data_rev_test.cpp test/unit/math/rev/mat/functor/map_rect_mpi_test.cpp test/unit/math/gpu/opencl_context_test.cpp
```

2. Check that `test-headers` works. If you're impatient:
```
make stan/math/prim/scal/fun/abs.hpp-test
```

3. Check that one of the prob tests work. This will not work on Windows.
```
./runTests.py test/prob/normal/normal_00000_generated_v_test.cpp
```


## Side Effects

The upstream tests will fail: both Stan and CmdStan rely on these makefiles. I will update those appropriately, but after we're ok with the proposed changes here.

Before merging, we need to update the Math wiki for building with MPI and GPU because flags would have changed.


## Additional Notes
Thanks to everyone helping out!

## Final TODOs

- [x] Update math wiki with instructions on build: https://github.com/stan-dev/math/wiki/Developer-Doc#building-and-running-tests
- [x] Update stan-dev/stan repo with new build. PR: stan-dev/stan#2601.
- [x] Update stan wiki with instructions on build: wiki updated: https://github.com/stan-dev/stan/wiki/Testing-Stan-using-Gnu-Make-and-Python#1-prerequisite-tools-and-settings
- [x] Update stan-dev/cmdstan repo with new build. PR: stan-dev/cmdstan#645 
- [x] Update cmdstan doc with instructions on build: updated tex.
